sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala

```diff
@@ -668,13 +668,37 @@ case class StreamingSymmetricHashJoinExec(
       private val iteratorNotEmpty: Boolean = super.hasNext
 
       override def completion(): Unit = {
-        val isLeftSemiWithMatch =
-          joinType == LeftSemi && joinSide == LeftSide && iteratorNotEmpty
-        // Add to state store only if both removal predicates do not match,
-        // and the row is not matched for left side of left semi join.
-        val shouldAddToState =
-          !stateKeyWatermarkPredicateFunc(key) && !stateValueWatermarkPredicateFunc(thisRow) &&
-          !isLeftSemiWithMatch
+        // The criteria of whether the input has to be added into state store or not:
+        // - Left side: input can be skipped to be added to the state store if it's already matched
+        //   and the join type is left semi.
+        //   For other cases, the input should be added, including the case it's going to be evicted
+        //   in this batch. It hasn't yet evaluated with inputs from right side "for this batch".
+        //   Learn about how the each side figures out the matches from other side.
```

Comment:
I don't understand what this line means. Can you clarify?

Reply:
We do not build a hashmap separately to match both sides and completely rely on state store. The way each side figures out the match is, looking into state store of other side. And there is a sequence of doing this, left side, and then right side.

That said, when the operator seeks for match from left side, right side is yet to be handled, hence left side can only see the right side for "prior batches". We take care of match in current batch from right side, assuming that we put the input of left side into state store. If we skip adding the input into state store for left side, we are missing possible matches rows (correctness issue).

Reply:
If the classdoc covers this I should just say read through classdoc. Do you get how it works from reading through classdoc?

Reply:
I understand the semantics, but I just didn't understand what your comment meant by "Learn about how...". Your latest commit addresses my concern.